File: flutter_app/lib/presentation/client/find_caregiver_view_model.dart
import 'package:flutter/material.dart';
import 'package:firebase_auth/firebase_auth.dart' as auth;
import 'package:cloud_firestore/cloud_firestore.dart';
import '../../domain/model/data_models.dart';

class FindCaregiverUiState {
  final bool isLoading;
  final List<User> caregivers;
  final String? errorMessage;

  FindCaregiverUiState({
    this.isLoading = true,
    this.caregivers = const [],
    this.errorMessage,
  });

  FindCaregiverUiState copyWith({
    bool? isLoading,
    List<User>? caregivers,
    String? errorMessage,
  }) {
    return FindCaregiverUiState(
      isLoading: isLoading ?? this.isLoading,
      caregivers: caregivers ?? this.caregivers,
      errorMessage: errorMessage ?? this.errorMessage,
    );
  }
}

class FindCaregiverViewModel extends ChangeNotifier {
  final auth.FirebaseAuth _auth = auth.FirebaseAuth.instance;
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  FindCaregiverUiState _uiState = FindCaregiverUiState();
  FindCaregiverUiState get uiState => _uiState;

  List<User> _allCaregivers = [];

  String _searchQuery = "";
  String get searchQuery => _searchQuery;

  String _selectedCategory = "All";
  String get selectedCategory => _selectedCategory;

  List<Review> _reviewsState = [];
  List<Review> get reviewsState => _reviewsState;

  bool _isReviewsLoading = false;
  bool get isReviewsLoading => _isReviewsLoading;

  List<String> _favoriteCaregiverIds = [];
  List<String> get favoriteCaregiverIds => _favoriteCaregiverIds;

  FindCaregiverViewModel() {
    fetchCaregiversFromFirebase();
    fetchFavoriteCaregivers();
  }

  void _setUiState(FindCaregiverUiState state) {
    _uiState = state;
    notifyListeners();
  }

  void fetchFavoriteCaregivers() {
    final currentUserId = _auth.currentUser?.uid;
    if (currentUserId == null) return;
    _firestore.collection("users").doc(currentUserId).snapshots().listen((snapshot) {
      if (snapshot.exists) {
        final user = User.fromMap(snapshot.data()!);
        _favoriteCaregiverIds = user.favoriteCaregivers;
        notifyListeners();
      }
    });
  }

  Future<void> toggleFavorite(String caregiverId) async {
    final currentUserId = _auth.currentUser?.uid;
    if (currentUserId == null) return;
    
    final currentFavs = List<String>.from(_favoriteCaregiverIds);
    if (currentFavs.contains(caregiverId)) {
      currentFavs.remove(caregiverId);
    } else {
      currentFavs.add(caregiverId);
    }
    
    await _firestore.collection("users").doc(currentUserId).update({
      "favoriteCaregivers": currentFavs
    });
  }

  Future<void> fetchCaregiversFromFirebase() async {
    try {
      final snapshot = await _firestore.collection("users")
          .where("role", whereIn: ["CAREGIVER", "NURSE"])
          .get();

      final todayStr = "${DateTime.now().year}-${DateTime.now().month.toString().padLeft(2, '0')}-${DateTime.now().day.toString().padLeft(2, '0')}";

      _allCaregivers = snapshot.docs
          .map((doc) => User.fromMap(doc.data()))
          .where((it) => (it.isVerified || it.kycStatus == "APPROVED") && !it.isBusy && !it.onLeaveDates.contains(todayStr))
          .toList();

      _allCaregivers.sort((a, b) => b.rating.compareTo(a.rating));
      _applyFilters();

    } catch (e) {
      _setUiState(_uiState.copyWith(isLoading: false, errorMessage: e.toString()));
    }
  }

  void updateSearchQuery(String query) {
    _searchQuery = query;
    _applyFilters();
  }

  void updateCategory(String category) {
    _selectedCategory = category;
    _applyFilters();
  }

  void _applyFilters() {
    var filteredList = _allCaregivers;

    if (_searchQuery.trim().isNotEmpty) {
      final query = _searchQuery.trim().toLowerCase();
      filteredList = filteredList.where((it) =>
          it.name.toLowerCase().contains(query) ||
          it.address.toLowerCase().contains(query)).toList();
    }

    if (_selectedCategory == "Caregiver") {
      filteredList = filteredList.where((it) => it.role == "CAREGIVER").toList();
    } else if (_selectedCategory == "Nurse") {
      filteredList = filteredList.where((it) => it.role == "NURSE").toList();
    }

    _setUiState(_uiState.copyWith(isLoading: false, caregivers: filteredList));
  }

  Future<void> fetchReviewsForCaregiver(String caregiverId) async {
    _isReviewsLoading = true;
    notifyListeners();
    try {
      final snapshot = await _firestore.collection("reviews")
          .where("caregiverId", isEqualTo: caregiverId)
          .get();

      _reviewsState = snapshot.docs.map((doc) => Review.fromMap(doc.data())).toList();
      _reviewsState.sort((a, b) => b.timestamp.compareTo(a.timestamp));
    } catch (e) {
      _reviewsState = [];
    } finally {
      _isReviewsLoading = false;
      notifyListeners();
    }
  }

  void clearReviews() {
    _reviewsState = [];
    notifyListeners();
  }
}

File: app/src/main/java/com/kina/care/presentation/client/FindCaregiverViewModel.kt
package com.kina.care.presentation.client

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.kina.care.domain.model.Review
import com.kina.care.domain.model.User
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

data class FindCaregiverUiState(
    val isLoading: Boolean = true,
    val caregivers: List<User> = emptyList(),
    val errorMessage: String? = null
)

class FindCaregiverViewModel : ViewModel() {
    private val firestore = FirebaseFirestore.getInstance()
    private val auth = FirebaseAuth.getInstance()
    private val currentUserId = auth.currentUser?.uid

    private val _uiState = MutableStateFlow(FindCaregiverUiState())
    val uiState: StateFlow<FindCaregiverUiState> = _uiState.asStateFlow()

    private var allCaregivers: List<User> = emptyList()

    val searchQuery = MutableStateFlow("")
    val selectedCategory = MutableStateFlow("All")

    private val _reviewsState = MutableStateFlow<List<Review>>(emptyList())
    val reviewsState: StateFlow<List<Review>> = _reviewsState.asStateFlow()

    var isReviewsLoading = MutableStateFlow(false)

    private val _favoriteCaregiverIds = MutableStateFlow<List<String>>(emptyList())
    val favoriteCaregiverIds: StateFlow<List<String>> = _favoriteCaregiverIds.asStateFlow()

    init {
        fetchCaregiversFromFirebase()
        fetchFavoriteCaregivers()
    }

    private fun fetchFavoriteCaregivers() {
        if (currentUserId == null) return
        firestore.collection("users").document(currentUserId)
            .addSnapshotListener { snapshot, _ ->
                if (snapshot != null && snapshot.exists()) {
                    val user = snapshot.toObject(User::class.java)
                    _favoriteCaregiverIds.value = user?.favoriteCaregivers ?: emptyList()
                }
            }
    }

    fun toggleFavorite(caregiverId: String) {
        if (currentUserId == null) return
        val currentFavs = _favoriteCaregiverIds.value.toMutableList()
        if (currentFavs.contains(caregiverId)) {
            currentFavs.remove(caregiverId)
        } else {
            currentFavs.add(caregiverId)
        }
        firestore.collection("users").document(currentUserId)
            .update("favoriteCaregivers", currentFavs)
    }

    private fun fetchCaregiversFromFirebase() {
        viewModelScope.launch {
            try {
                val snapshot = firestore.collection("users")
                    .whereIn("role", listOf("CAREGIVER", "NURSE"))
                    .get()
                    .await()

                val todayStr = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(Date())

                val list = snapshot.documents.mapNotNull { it.toObject(User::class.java) }
                    // 🌟 Fix: kycStatus == "APPROVED" වූ විටත් ගිණුම් පෙන්වයි
                    .filter { (it.isVerified || it.kycStatus == "APPROVED") && !it.isBusy && !it.onLeaveDates.contains(todayStr) }
                    .sortedByDescending { it.rating }

                allCaregivers = list
                applyFilters()

            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(isLoading = false, errorMessage = e.localizedMessage)
            }
        }
    }

    fun updateSearchQuery(query: String) {
        searchQuery.value = query
        applyFilters()
    }

    fun updateCategory(category: String) {
        selectedCategory.value = category
        applyFilters()
    }

    private fun applyFilters() {
        var filteredList = allCaregivers

        val query = searchQuery.value.trim()
        if (query.isNotEmpty()) {
            filteredList = filteredList.filter {
                it.name.contains(query, ignoreCase = true) ||
                        it.address.contains(query, ignoreCase = true)
            }
        }

        // 🌟 Fix: Caregiver සහ Nurse ලෙස Role එක අනුව Filter කිරීම
        val category = selectedCategory.value
        if (category == "Caregiver") {
            filteredList = filteredList.filter { it.role == "CAREGIVER" }
        } else if (category == "Nurse") {
            filteredList = filteredList.filter { it.role == "NURSE" }
        }

        _uiState.value = _uiState.value.copy(isLoading = false, caregivers = filteredList)
    }

    fun fetchReviewsForCaregiver(caregiverId: String) {
        viewModelScope.launch {
            isReviewsLoading.value = true
            try {
                val snapshot = firestore.collection("reviews")
                    .whereEqualTo("caregiverId", caregiverId)
                    .get()
                    .await()

                val reviews = snapshot.documents.mapNotNull { it.toObject(Review::class.java) }
                    .sortedByDescending { it.timestamp }

                _reviewsState.value = reviews
            } catch (e: Exception) {
                e.printStackTrace()
                _reviewsState.value = emptyList()
            } finally {
                isReviewsLoading.value = false
            }
        }
    }

    fun clearReviews() {
        _reviewsState.value = emptyList()
    }
}